group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -761,6 +777,31 @@ public ClassicGroup classicGroup(
         }
     }
 
+    public boolean validateOnlineUpgrade(ClassicGroup classicGroup) {
+        return ConsumerGroupMigrationPolicy.isUpgradeEnabled(consumerGroupMigrationPolicy) &&
+            !classicGroup.isInState(DEAD) &&
+            ConsumerProtocol.PROTOCOL_TYPE.equals(classicGroup.protocolType().orElse(null)) &&
+            classicGroup.size() <= consumerGroupMaxSize;
+    }
+
+    ConsumerGroup convertToConsumerGroup(ClassicGroup classicGroup, List<Record> records) {
+        classicGroup.completeAllJoinFutures(Errors.REBALANCE_IN_PROGRESS);
+        classicGroup.completeAllSyncFutures(Errors.REBALANCE_IN_PROGRESS);
+
+        createGroupTombstoneRecords(classicGroup, records);
```

Comment:
nit: You could directly use `classicGroup.createGroupTombstoneRecords(records);`.